### PR TITLE
Add the ability to be damaged by players (melee)

### PR DESCRIPTION
**Pull request by turt2live**
This PR adds the ability for players to be melee attacked by other players. This does not handle things like potions, arrows, etc. Specified non-standard weapons are damaged upon impact as well.

Included is a mapping of item type to damage "force". The same utility class can also critical hits for whoever the poor soul that does that is. 

Some additional utility methods were added for future support beyond damage, but they are directly used by this PR. 

---

_Edited by turt2live_

**Related Links:**
- Ticket: #203

### CODE REVIEW

**Comment by dequis** [comment]
Good job @turt2live and @turt2bot

File: src/main/java/net/glowstone/constants/ItemDamage.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                case IRON_PICKAXE:
+                case DIAMOND_PICKAXE:
+                case WOOD_SPADE:
+                case GOLD_SPADE:
+                case IRON_SPADE:
+                case DIAMOND_SPADE:
+                    item.setDurability((short) (item.getDurability() + 2));
+                    break;
+                default:
+                    break;
```
Comment (dequis) [inline src/main/java/net/glowstone/constants/ItemDamage.java]: Normal swords lose 1 durability on successful hits too

File: src/main/java/net/glowstone/constants/AttackDamage.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            case GOLD_SWORD:
+            case IRON_SWORD:
+            case DIAMOND_SWORD:
+                return 1;
+            default:
+                return 0;
+        }
+    }
+
+    // TODO: Other methods for stuff like lava contact, fire, arrows, etc
```
Comment (dequis) [inline src/main/java/net/glowstone/constants/AttackDamage.java]: Irrelevant since this is about attack damage now, right?

Comment (turt2live) [inline src/main/java/net/glowstone/constants/AttackDamage.java]: Some forms of damage not implemented in this PR are still applicable. However they can be added on a case-by-case basis. (removing line)

File: src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java [hunk 3]
```diff
@@ -1,12 +1,77 @@
...
+            GlowServer.logger.info("Player " + player + " tried to interact with an entity while dead");
+            return;
+        }
+
+        GlowEntity possibleTarget = player.getWorld().getEntityById(message.getId());
+        GlowLivingEntity target = possibleTarget instanceof GlowLivingEntity ? (GlowLivingEntity) possibleTarget : null;
+
+        if (message.getAction() == InteractEntityMessage.Action.ATTACK.ordinal()) {
+            if (target == null) {
+                GlowServer.logger.info("Player " + player + " tried to attack an entity that does not exist");
```
Comment (dequis) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java:41]: warning instead?

Comment (turt2live) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java:41]: I personally don't care. I'll wait for feedback

Comment (bendem) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java:41]: Same line 32.

**Comment by mamifsidtect** [comment]
Dat branch name tho, but yeah, I would also suggest putting a warning instead of a log

File: src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java [hunk 4]
```diff
@@ -1,12 +1,77 @@
...
+                            hand.setDurability((short) (hand.getDurability() + durabilityLoss));
+                        }
+                        player.updateInventory();
+                    }
+                }
+            }
+        } else if (message.getAction() == InteractEntityMessage.Action.ATTACK_AT.ordinal()) {
+            // TODO: Interaction with entity (right click)
+        } else if (message.getAction() == InteractEntityMessage.Action.INTERACT.ordinal()) {
+            // TODO: Handle something (unknown cause for this action)
```
Comment (SpaceManiac) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java]: Comments here seem to mismatch the branch headers. I understand `INTERACT` corresponds to a right-click and `ATTACK_AT` is some specialty armor stand thing. Is this incorrect?

File: src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java (hunk 3 above)
Comment (SpaceManiac) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java:41]: As-is is probably fine in this case.

File: src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java (hunk 4 above)
Comment (turt2live) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java]: Due to the use of ordinal (may be my fault?) the enum names are incorrect for the branches. Actual testing produced the comments.

Comment (gabizou) [inline src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java]: I'm fairly certain we should then change the order of the Action enum? I mean, we can't have these types of things confusing devs, now can we :P

**Comment by gdude2002** [comment]
Builds fine here, complete with lyrics. http://bamboo.gserv.me/browse/GLOW-SRV-PR461-125/log

**Comment by SpaceManiac** [comment]
Merged with tweaks in 39ca0abb5ae26c919b991724e02a31c854414d03. Thanks!

**Pull request closed without merging**